File: medium/3635_earliest_finish_time_for_land_and_water_rides_ii/solution.py
from typing import List


class Solution:
    def earliestFinishTime(
        self,
        landStartTime: List[int],
        landDuration: List[int],
        waterStartTime: List[int],
        waterDuration: List[int],
    ) -> int:
        earliestLandFinishTime = float("inf")
        for i in range(len(landStartTime)):
            if landStartTime[i] >= earliestLandFinishTime:
                continue

            earliestLandFinishTime = min(
                earliestLandFinishTime, landStartTime[i] + landDuration[i]
            )

        earliestWaterFinishTime = float("inf")
        for i in range(len(waterStartTime)):
            if waterStartTime[i] >= earliestLandFinishTime:
                continue

            earliestWaterFinishTime = min(
                earliestWaterFinishTime, waterStartTime[i] + waterDuration[i]
            )

        earliestFinishTime = float("inf")

        for i in range(len(landStartTime)):
            earliestFinishTime = min(
                earliestFinishTime,
                max(earliestWaterFinishTime, landStartTime[i])
                + landDuration[i],
            )

        for i in range(len(waterStartTime)):
            earliestFinishTime = min(
                earliestFinishTime,
                max(earliestLandFinishTime, waterStartTime[i])
                + waterDuration[i],
            )

        return earliestFinishTime

File: medium/3635_earliest_finish_time_for_land_and_water_rides_ii/Solution.java
class Solution {
  public int earliestFinishTime(
      int[] landStartTime, int[] landDuration, int[] waterStartTime, int[] waterDuration) {
    int earliestLandFinishTime = Integer.MAX_VALUE;
    for (int i = 0; i < landStartTime.length; ++i) {
      earliestLandFinishTime = Math.min(earliestLandFinishTime, landStartTime[i] + landDuration[i]);
    }

    int earliestWaterFinishTime = Integer.MAX_VALUE;
    int earliestFinishTime = Integer.MAX_VALUE;

    for (int i = 0; i < waterStartTime.length; ++i) {
      earliestWaterFinishTime =
          Math.min(earliestWaterFinishTime, waterStartTime[i] + waterDuration[i]);
      earliestFinishTime =
          Math.min(
              earliestFinishTime,
              Math.max(earliestLandFinishTime, waterStartTime[i]) + waterDuration[i]);
    }

    for (int i = 0; i < landStartTime.length; ++i) {
      earliestFinishTime =
          Math.min(
              earliestFinishTime,
              Math.max(earliestWaterFinishTime, landStartTime[i]) + landDuration[i]);
    }

    return earliestFinishTime;
  }
}
